### ISSUE TITLE
How to score the likelihoods across a DNA sequence with web page or api?

### ISSUE THREAD

**Issue by shengs-UCAS** [ArcInstitute/evo2#77]
It seems that sequences can only be generated on web pages https://build.nvidia.com/arc/evo2-40b, how to get the likelihoods/probability values instead of images as https://arcinstitute.org/tools/evo/evo-designer

<img width="922" alt="Image" src="https://github.com/user-attachments/assets/79ca03ea-e693-4533-9da6-a01d5f5fa71b" />

**Comment by curiosusJR** [comment]
I have the same question. I can get the logits followed README, but how can I convert it to this sequence score (likelihood)?

**Comment by LauferVA** [comment]
@garykbrixi  - I am still getting up to speed on this package, but I think this may close #77 .. your feedback most welcome.

---------------

Hi @shengs-UCAS  and @curiosusJR, 

The conversion from logits to sequence likelihood is handled in `scoring.py` by applying `torch.log_softmax` to the logits (via the `logits_to_logprobs` function) and then gathering the log probabilities for the actual token IDs. These values are then aggregated (either summed for the PLL or averaged) in the helper function `_score_sequences`.

Ok, now then. @curiosusJR - it seems that you have been successful in getting the logits as is described in the README. If you have these,  **you can compute the sequence score directly by applying a log softmax and reducing the resulting probabilities**.  To do this, you could call the functions from the scoring module directly, on your own computer. One would import and call the helper function `logits_to_logprobs` from `scoring.py`. That function takes the raw logits, applies a log softmax, and gathers the token log probabilities using the provided input IDs.

**Quick answer - use the helper function to get likelihoods if you already have logits**:
```python
from evo2.scoring import logits_to_logprobs
import torch
import torch.nn.functional as F

# Suppose these are your model outputs and inputs:
batch_size = 2
seq_length = 10
vocab_size = 10000

# Example logits and input_ids:
logits = torch.randn(batch_size, seq_length, vocab_size)
input_ids = torch.randint(0, vocab_size, (batch_size, seq_length))

# Compute log probabilities for the actual tokens:
log_probs = logits_to_logprobs(logits, input_ids)
print("Log probabilities:", log_probs)
```

**A more general/powerful workflow:**
However, let's say that, instead,  you interested in scoring full sequences, or you want batch handling, or you want reverse-complement processing, etc., etc...). 

There is a more flexible way forward: you can call the higher-level function `score_sequences` (or `score_sequences_rc`) directly, which internally uses `logits_to_logprobs` and aggregates the scores. So, if you’re working with sequences and a model/ tokenizer, you might prefer:

```python
from evo2.scoring import score_sequences

sequences = ["ACGTACGT", "TGCAATGC"]
# Assuming you have a model instance (of type StripedHyena) and a tokenizer, e.g.,
# model = <your evo2 model instance>
# tokenizer = <your tokenizer instance>
scores = score_sequences(sequences, model=model, tokenizer=tokenizer)
print("Sequence scores:", scores)
```

So yes—the functions in `scoring.py` can be used directly to convert logits into sequence likelihoods.

**Comment by curiosusJR** [comment]
thank you for your timely reply! Now I get it.

**Comment by garykbrixi** [comment]
Thank you @LauferVA, closing this issue now

**Comment by shengs-UCAS** [comment]
@garykbrixi @LauferVA  Thanks a lot for the timely reply! It seems to be a good solution if getting the logits. 
Due to hardware limitations (3090), I am unable to set it up locally. 
What should I do to get the sequence likelihoods on the NVIDIA NIM (https://health.api.nvidia.com/v1/biology/arc/evo2-40b/forward) or in other ways?
